python/repositories.bzl

```diff
@@ -1,8 +1,225 @@
+    "PLATFORMS",
+    "TOOL_VERSIONS",
+    "get_release_url",
+)
+
 def py_repositories():
     # buildifier: disable=print
     print("py_repositories is a no-op and is deprecated. You can remove this from your WORKSPACE file")
+
+########
+# Remaining content of the file is only used to support toolchains.
+########
+
+def _python_repository_impl(rctx):
+    platform = rctx.attr.platform
+    python_version = rctx.attr.python_version
+    (release_filename, url) = get_release_url(platform, python_version)
+
+    if release_filename.endswith(".zst"):
+        rctx.download(
+            url = url,
+            sha256 = rctx.attr.sha256,
+            output = release_filename,
+        )
+        unzstd = rctx.which("unzstd")
+        if not unzstd:
+            url = rctx.attr.zstd_url.format(version = rctx.attr.zstd_version)
+            rctx.download_and_extract(
+                url = url,
+                sha256 = rctx.attr.zstd_sha256,
+            )
+            working_directory = "zstd-{version}".format(version = rctx.attr.zstd_version)
+            rctx.execute(
+                ["make", "--jobs=4"],
+                timeout = 600,
+                quiet = True,
+                working_directory = working_directory,
+            )
+            zstd = "{working_directory}/zstd".format(working_directory = working_directory)
+            unzstd = "./unzstd"
+            rctx.symlink(zstd, unzstd)
+
+        exec_result = rctx.execute([
+            "tar",
+            "--extract",
+            "--strip-components=2",
+            "--use-compress-program={unzstd}".format(unzstd = unzstd),
+            "--file={}".format(release_filename),
+        ])
+        if exec_result.return_code:
+            fail(exec_result.stderr)
+    else:
+        rctx.download_and_extract(
+            url = url,
+            sha256 = rctx.attr.sha256,
+            stripPrefix = "python",
+        )
+
+    # Make the Python installation read-only.
+    if "windows" not in rctx.os.name:
+        exec_result = rctx.execute(["chmod", "-R", "ugo-w", "lib"])
+        if exec_result.return_code:
+            fail(exec_result.stderr)
+
+    python_bin = "python.exe" if ("windows" in rctx.attr.platform) else "bin/python3"
+
+    build_content = """\
+# Generated by python/repositories.bzl
+
+load("@bazel_tools//tools/python:toolchain.bzl", "py_runtime_pair")
+
+package(default_visibility = ["//visibility:public"])
+
+filegroup(
+    name = "files",
+    srcs = glob(
+        include = [
+            "bin/**",
+            "DLLs/**",
+            "extensions/**",
+            "include/**",
+            "lib/**",
+            "libs/**",
+            "Scripts/**",
+            "share/**",
+        ],
+        exclude = [
+            "**/* *", # Bazel does not support spaces in file names.
+        ],
+    ),
+)
+
+exports_files(["{python_path}"])
+
+py_runtime(
+    name = "py2_runtime",
+    interpreter_path = "/bin/false",
+    python_version = "PY2",
+)
```

Comment:
What's the reason for defining this? 
It means checks like `if toolchain.py2_runtime:` will always be true, which seems odd given that it points to `/bin/false`.
Could the `py2_runtime` be unset?

Reply:
Addressed in f509b46.